lib/SILOptimizer/Differentiation/JVPCloner.cpp

```diff
@@ -1475,18 +1532,41 @@ void JVPCloner::Implementation::prepareForDifferentialGeneration() {
   // Initialize tangent mapping for indirect results.
   auto origIndResults = original->getIndirectResults();
   auto diffIndResults = differential.getIndirectResults();
+  auto isNonWrtInoutParameter = [&](unsigned i) {
+    auto &paramInfo = original->getLoweredFunctionType()->getParameters()[i];
+    return paramInfo.isIndirectInOut() && !getIndices().parameters->contains(i);
+  };
+  auto origNumParams = original->getLoweredFunctionType()->getNumParameters();
 #ifndef NDEBUG
-  unsigned numNonWrtInoutParameters = llvm::count_if(
-    range(original->getLoweredFunctionType()->getNumParameters()),
-    [&] (unsigned i) {
-      auto &paramInfo = original->getLoweredFunctionType()->getParameters()[i];
-      return paramInfo.isIndirectInOut() && !getIndices().parameters->contains(i);
-    });
+  unsigned numNonWrtInoutParameters =
+      llvm::count_if(range(origNumParams), isNonWrtInoutParameter);
+  assert(origIndResults.size() + numNonWrtInoutParameters ==
+         diffIndResults.size());
 #endif
-  assert(origIndResults.size() + numNonWrtInoutParameters == diffIndResults.size());
+  differentialBuilder.setInsertionPoint(differential.getEntryBlock());
+  unsigned j = 0;
+  for (auto i : range(origNumParams)) {
+    // Non-wrt inout parameters become indirect results of the differential.
+    if (isNonWrtInoutParameter(i)) {
+      auto &origParam = original->getArgumentsWithoutIndirectResults()[i];
+      auto &tanBuf = diffIndResults[j + origIndResults.size()];
+      setTangentBuffer(origEntry, origParam, tanBuf);
+      // Original inout parameters are initialized, therefore the differential
+      // indirect result must be initialized too.
+      emitZeroIndirect(tanBuf->getType().getASTType(), tanBuf,
+                       differential.getLocation());
+    }
+  }
   for (auto &origBB : *original)
     for (auto i : indices(origIndResults))
       setTangentBuffer(&origBB, origIndResults[i], diffIndResults[i]);
+
+  // Tangent buffers for non-varied original indirect results have to be
+  // initialized.
+  for (auto i : indices(origIndResults))
+    if (!activityInfo.isVaried(origIndResults[i], getIndices().parameters))
+      emitZeroIndirect(diffIndResults[i]->getType().getASTType(),
+                       diffIndResults[i], differential.getLocation());
```

Comment:
Rather than directly iterating over original indirect results, we should iterate over differentiability result indices. This avoids including original indirect results that are not relevant for differentiation.

Reply:
Fixed in https://github.com/apple/swift/pull/33633/commits/d37e2f46c047777814e4a3f92b20907fd5795969.